lib/client/client.dart

```diff
@@ -0,0 +1,159 @@
+import 'dart:async';
+import 'dart:ffi';
+import 'dart:io';
+import 'dart:isolate';
+import 'dart:typed_data';
+
+import 'package:ffi/ffi.dart';
+import 'package:path/path.dart' as path;
+
+import 'native_client.dart';
+
+class SendResult {
+  String code;
+  Future<void> done;
+
+  SendResult(this.code, this.done);
+}
+
+class Client {
+  late NativeClient _native;
+  late Config? _config;
+
+  Client({Config? config}) {
+    _native = NativeClient(config: config);
+    _config = config;
+  }
+
+  Future<SendResult> sendText(String msg) {
+    final done = Completer<void>();
+    // TODO: much much is it allocating?
+    Pointer<Pointer<Utf8>> codeOut = calloc();
+
+    final rxPort = ReceivePort()
+      ..listen((dynamic errCode) {
+        if (errCode is int) {
+          // TODO: Create exception implementation(s).
+          final exception =
+              Exception('Failed to send text. Error code: $errCode');
+          done.completeError(exception);
+          throw exception;
+        }
+        done.complete();
+      });
+
+    final errCode = _native.clientSendText(
+        msg.toNativeUtf8(), codeOut, rxPort.sendPort.nativePort);
+
+    if (errCode != 0) {
+      // TODO: Create exception implementation(s).
+      calloc.free(codeOut);
+      final exception = Exception('Failed to send text. Error code: $errCode');
+      done.completeError(exception);
+      throw exception;
+    }
+
+    final Pointer<Utf8> code = codeOut.value;
+    calloc.free(codeOut);
+
+    final result = SendResult(code.toDartString(), done.future);
+    return Future.value(result);
+  }
+
+  Future<String> recvText(String code) {
+    final done = Completer<String>();
+
+    final rxPort = ReceivePort()
+      ..listen((dynamic response) {
+        if (response is int) {
+          // TODO: Create exception implementation(s).
+          final exception =
+              Exception('Failed while sending text. Error code: $response');
+          done.completeError(exception);
+          throw exception;
+        }
+
+        done.complete(String.fromCharCodes(response));
+      });
+
+    final int errCode =
+        _native.clientRecvText(code.toNativeUtf8(), rxPort.sendPort.nativePort);
+    if (errCode != 0) {
+      // TODO: Create exception implementation(s).
+      throw Exception('Failed to send text. Error code: $errCode');
+    }
+
+    return done.future;
+  }
+
+  Future<SendResult> sendFile(File file) async {
+    final fileName = path.basename(file.path);
+    final length = await file.length();
+    print("File length was: $length");
+    final done = Completer<void>();
+
+    Pointer<Pointer<Utf8>> codeOut = calloc();
```

Comment:
We should look into this, what is `calloc` doing exactly?

Reply:
To reiterate our joint digging for this one, `calloc` is an `Allocator` and allocators are callable due to an extension function defined on `Allocator` that takes an optional count(default is 1).